keps/sig-auth/4872-harden-kubelet-cert-validation/README.md

```diff
@@ -0,0 +1,389 @@
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+- [x] Metrics
+  - Metric name: `kube_apiserver_pod_logs_backend_tls_failure_total`
+  - Components exposing the metric: kube-apiserver
+
+> TODO: should `kube_apiserver_pod_logs_backend_tls_failure_total` reflect errors due to the new CN validation?
+> It's technically a TLS failure, but it's not part of the base TLS client validations.
+
+###### Are there any missing metrics that would be useful to have to improve observability of this feature?
+
+We could add a metric specific to track the number of requests that failed due to the new CN validation. In addition, we could track the time spent per request on the CN validation.
+
+However, we consider these metrics to not provide enough value to justify the work to maintain them.
+
+### Dependencies
+
+###### Does this feature depend on any specific services running in the cluster?
+
+No.
+
+### Scalability
+
+###### Will enabling / using this feature result in any new API calls?
+
+No.
+
+###### Will enabling / using this feature result in introducing new API types?
+
+No.
+
+###### Will enabling / using this feature result in any new calls to the cloud provider?
+
+No.
+
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+No.
+
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+No. This only affects streaming APIs and these are not covered by SLIs/SLOs.
+
+###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
+
+No.
+
+Note: depending on the implementation (caching the client-go transport or not) there might be a slight increase in memory (due to one transport per node being cached) or in CPU usage (due to building the transport on the fly for every request). This should be negligible.
+
+###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?
+
+No.
+
+### Troubleshooting
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+It's part of the API server, so the feature will be unavailable.
+
+###### What are other known failure modes?
+
+- [API server can't connect to Nodes with custom kubelet serving certificates that don't follow the `system:node:<node-name>` convention]
+  - Detection: `kubectl logs` returns a certificate validation error. 
+  - Mitigations: disable the validation with the `--disable-kubelet-cert-cn-validation` flag.
+  - Diagnostics: error is returned by the API server, no additional logging needed.
+  - Testing: We will have tests for this, this is basically testing that the feature works. 
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
```

Comment:
Add a link to https://docs.google.com/document/d/1RqhAkGov_coHsB3lbAo-qfQl1MOfYvgpPUjiGMJ_3PY here since that discussion helped narrow down the set of implementation choices we have.